plugins/pulp_rpm/plugins/distributors/yum/metadata/updateinfo.py

```diff
@@ -13,12 +16,62 @@
                                           UPDATE_INFO_XML_FILE_NAME)
+        self.conduit = conduit
         super(UpdateinfoXMLFileContext, self).__init__(
             metadata_file_path, 'updates', checksum_type=checksum_type)
 
+    def _repo_unit_nevra(self, erratum_unit, repo_id):
+        """
+        Return a list of NEVRA dicts for units in a single repo referenced by the given errata.
+
+        Pulp errata units combine the known packages from all synced repos. Given an errata unit
+        and a repo, return a list of NEVRA dicts that can be used to filter out packages not
+        linked to that repo when generating a repo's updateinfo XML file. While returning that
+        list of NEVRA dicts is the main goal, doing so quickly and without running out of memory
+        is what makes this a little bit tricky.
+
+        Build up a super-fancy query to get the unit ids for all NEVRA seen in these errata
+        check repo/unit associations for this errata to limit the packages in the published
+        updateinfo to the units in the repo being currently published.
+
+        :param erratum_unit: The erratum unit that should be written to updateinfo.xml.
+        :type erratum_unit: pulp_rpm.plugins.db.models.Errata
+        :param repo_id: The repo_id of a pulp repository in which to find units
+        :type repo_id: str
+        """
+        nevra_fields = ('name', 'epoch', 'version', 'release', 'arch')
+        nevra_q = mongoengine.Q()
+        for pkglist in erratum_unit.pkglist:
+            for pkg in pkglist['packages']:
+                pkg_nevra = dict((field, pkg[field]) for field in nevra_fields)
+                nevra_q |= mongoengine.Q(**pkg_nevra)
+
+        # Aim the super-fancy query at mongo to get the units that this errata refers to
+        # The scaler method on the end returns a list of tuples to try to save some memory
+        # and also cut down on mongoengine model instance hydration costs.
+        nevra_units = models.RPM.objects.filter(nevra_q).scalar('id', *nevra_fields)
+
+        # Split up the nevra unit entries into a mapping of the unit id to its nevra fields
+        nevra_unit_map = dict((nevra_unit[0], nevra_unit[1:]) for nevra_unit in nevra_units)
+
+        # Get all of the unit ids from this errata that are associated with the current repo.
+        repo_unit_ids = RepositoryContentUnit.objects.filter(
+            unit_id__in=nevra_unit_map.keys(), repo_id=repo_id).scalar('unit_id')
+
+        # Finally(!), intersect the repo unit ids with the unit nevra ids to
+        # create a list of nevra dicts that can be easily compared to the
+        # errata package nevra and exclude unrelated packages
+        repo_unit_nevra = []
+        for nevra_unit_id, nevra_field_values in nevra_unit_map.items():
+            # based on the args to scalar when nevra_units was created:
+            if nevra_unit_id in repo_unit_ids:
```

Comment:
In case ``repo_unit_ids`` has many entries, it would perform this operation better as a set.